Destroy CUDA events after profiling

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39962 Destroy CUDA events after profiling**

Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

Commit:
Destroy CUDA events after profiling

Summary:
Adding a simple wrapper with ref count for cuda event and
destroying cuda event after the last copy is destroyed

Test Plan:
CI cuda profiler tests

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Adding a simple wrapper with ref count for cuda event and
destroying cuda event after the last copy is destroyed

Test Plan:
CI cuda profiler tests

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Adding a simple wrapper with ref count for cuda event and
destroying cuda event after the last copy is destroyed

Test Plan:
CI cuda profiler tests

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Test Plan:
CI cuda profiler tests

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

ngimel commented:
Can you please add a test to make sure that events are not leaked, as in the original issue?

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]

Commit:
Update on "Destroy CUDA events after profiling"


Summary:
Using shared_ptr for cuda event and
destroying cuda event after the last copy is destroyed

Differential Revision: [D22027092](https://our.internmc.facebook.com/intern/diff/D22027092)

[ghstack-poisoned]
